### DISCUSSION TITLE
Tuya MCU driver testing - help wanted

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#15281] [General]
Help Wanted with TAS & Tuya testing. 

I've reworked the tuya driver, but only have a single device for testing. 
I've raised an [issue](https://github.com/btsimonh/Tasmota/issues/11) on my repo - where I've referenced my branch and attached a -lite binary. 
If anyone can help by testing it on other Tuya MCU products, please report/discuss there? 
I don't want to PR until it's seen some testing other than on my Dimmer.

Basic testing needed:
If you can get DEBUG logs from boot (e.g. using MQTT logs or syslog), then please do so, as this could highlight any issues.
(see issue above for first reports - it's already failing for some).

**Comment by btsimonh** [comment]
Hi All, 
sorry to bash on about Tuya Testing: new version in issue here: https://github.com/btsimonh/Tasmota/issues/11.  Changes relate to diagnosing WF-DS01.  I really need others to test and report - e.g. dual dimmer, sensors, etc.

Ref logs of issues, I log using an MQTT node in Node-Red appending to a file, subscribed to e.g. stat/ATDIM-1/# with MWTT logging turned up to the max.

br,
S

**Comment by btsimonh** [comment]
I've enabled the wiki on my repo specifically for Tuya testing info, and added testing guidance:
https://github.com/btsimonh/Tasmota/wiki/Testing-Strategies---Dimmers

**Comment by btsimonh** [comment]
Note firmware updates in wiki (https://github.com/btsimonh/Tasmota/wiki/Current-Binaries).
The mod is simple - ignore TAS trying to set dimmer values if the command was causes by the MCU.  This should avoid odd feedback loops with the MCU basically trying to control itself.
Any help with testing very much appreciated - these is a testing strategy doc in the wiki, and brief instructions on how to get a complete log.

**Comment by btsimonh** [comment]
drag this to the top again - please test, else no PR which may help you
I have some time out over the next 2-3 weeks.  If no solid responses, I'll issue a PR when I am back,

**Comment by btsimonh** [comment]
logging in binaries cut down - more cryptic but hopefully we can avoid some log loss.  Getting good logs seems to be the most challenging part of testing.  There is a description of how to get complete logs in the wiki in my repo, linked above.

**Comment by m-jarzebowski** [comment]
Do you still need someone to test WF-DS01?

**Comment by esackbauer** [comment]
I have also a WF-DS01, currently running only on 8.5.1. For me the reason is that only with this version the dimmer does "soft on".
I need this because otherwise the protection switch for my halogen bulb transformater is triggered (when switching on with 100% brightness).
Let me know if I shall test something.

**Reply by m-jarzebowski** [reply]
Here you can find some more info -> https://github.com/btsimonh/Tasmota/issues/11
Maybe it will be helpful.

**Comment by mvanlijden** [comment]
Sorry for not replying the request for testing, life came in between.
I will try to find some time tonight or other nights to test the binaries and post results in the branch. I can easily whip up a test setup separate from my built in lights :)

**Reply by btsimonh** [reply]
:) it's a good time to test - I have had a couple of months away from TAS with work & holiday, but thinking of adding a PR in the next couple of weeks;  of the people who have tested, everything has worked ok, except maybe restoring the last (manual?) dim value when doing OFF/ON.  The issue here is that some dimmers report all the dim values back to TAS as the light dims to dark - so TAS can't store the received DIM value, else it would restore '0' on the next ON.

**Reply by esackbauer** [reply]
Do you know if "Soft On" Function is working on WF-DS01 with your latest changes? It seems it only worked for me until 8.5.1, then all newer versions are instantly on.

**Reply by mvanlijden** [reply]
@btsimonh as posting in testing issue on your branch:

Can confirm that on my WF-DS01, with a simple philips 6W dimmable light bulb, all seems to working extremely well and fast in response. 

Tested and confirmed working:
- Direct command with random switching power, dimmer, power, dimmer, dimmer, incl backlog testing always ending on last value
- Buttons work with a very fast response, dimming only works when on (like designed i think). Always restoring last value.
- Switching and dimmer randomly with buttons, web, mqtt, web, button, etc. Have found no issues of missing a adjustment.
- Dimming value updated correctly in HA (so MQTT) when using physical buttons
- Behavior with SetOption20 and SetOption54 enabled works as designed, shifting to stored dim value 2 sec after power on

Cannot find any scenario that does not meet expectations. For me this feels much more stable and also responsive. Would love to see it merged into main :)

**Reply by esackbauer** [reply]
Does it do soft-on to the last value?

**Reply by mvanlijden** [reply]
> Does it do soft-on to the last value?

As far is I know soft-on was never a default function of Tasmota. Only the Soft-Off is handled by the MCU. Soft-on was implemented by Tuya in their firmware. Look at the Fade option (together with SetOption91) or the Wakeup and WakeupDuration if this works for you.

**Reply by esackbauer** [reply]
Hi Maarten, thanks for your reply. I was asking because until Tasmota Version  8.5.1 it would do soft-on, as it did with stock Tuya firmware.
So I guess it is dependend on the embedded TuyaMCU firmware of Tasmota?

**Comment by esackbauer** [comment]
Just wanted to thank you guys for providing the 11.0.0.5 version. It works really good and stable with my WF-DS01.
Also Soft-On and Soft-Off works great.

**Comment by mvanlijden** [comment]
Just a question, seeing that Tasmota is already on 12.0.2, will you need to rebase on that? 
I saw some updates on Tuya in the 11.1 releasenotes (https://github.com/arendst/Tasmota/pull/15267 and https://github.com/arendst/Tasmota/pull/15216)
Do you have a plan going forward?

**Comment by btsimonh** [comment]
hi @mvanlijden,
I think I already rebased on top of those changes, but would need to confirm.  I've not had much time for TAS recently, but the feedback on my mods seems good, so it's probably time to PR.  I'll see what I can do in the coming week, unless someone else offers to do it?
br,
Simon

**Comment by btsimonh** [comment]
I've put it on top of latest dev version, see https://github.com/btsimonh/Tasmota/issues/13
I'm unable to test in a real device for a while, but have built fat and lite versions and attached in the wiki.
If we get a couple of positive tests that it runs, and performs the same as the 11 version, I'll issue a PR for it so hopefully it will stay in TAS.
br,
Simon

**Comment by arendst** [comment]
See latest dev release v12.2.0.4 and #17051

**Comment by rktaiwala** [comment]
I have two tuya module WB3S and CB2S, I can test on those.

**Reply by btsimonh** [reply]
no, unfortunately those are not ESP based.  
For them you can go Openbeken, but not yet Tasmota.   
What we need testing is against serial Tuya MCUs in ESP8266 or ESP32 based devices running Tasmota.
Thankyou for commenting though :).

**Reply by ABBAGABBA** [reply]
> no, unfortunately those are not ESP based. For them you can go Openbeken, but not yet Tasmota. What we need testing is against serial Tuya MCUs in ESP8266 or ESP32 based devices running Tasmota. Thankyou for commenting though :).

Hello! Could you tell about driver modification process in detailes or give me link? I tried to migrate from cb3s to esp8266 on gsm alarm system pg107, but it dont work fully, alarm system dont see wifi state, time and etc.
